src/lib/reducer.ts

```diff
@@ -0,0 +1,184 @@
+import type { ComponentsObject, HttpMethods, OASDocument, TagObject } from 'rmoas.types';
+
+import jsonPath from 'jsonpath';
+import jsonPointer from 'jsonpointer';
+import { version as getAPIDefinitionVersion } from 'oas-normalize/dist/lib/utils';
+
+export type ReducerOptions = {
+  /** A key-value object of path + method combinations to reduce by. */
+  paths?: Record<string, string[] | '*'>;
+  /** An array of tags in the OpenAPI definition to reduce by. */
+  tags?: string[];
+};
+
+/**
+ * Query a JSON Schema object for any `$ref` pointers. Return any pointers that were found.
+ *
+ * @param schema JSON Schema object to look for any `$ref` pointers within it.
+ */
+function getUsedRefs(schema: any) {
+  return jsonPath.query(schema, "$..['$ref']");
+}
+
+/**
+ * Recursively process a `$ref` pointer and accumulate any other `$ref` pointers that it or its
+ * children use.
+ *
+ * @param schema JSON Schema object to look for and accumulate any `$ref` pointers that it may have.
+ * @param $refs Known set of `$ref` pointers.
+ * @param $ref `$ref` pointer to fetch a schema from out of the supplied schema.
+ */
+function accumulateUsedRefs(schema: Record<string, unknown>, $refs: Set<string>, $ref: string): void {
+  const $refSchema = jsonPointer.get(schema, $ref.substring(1));
+  getUsedRefs($refSchema).forEach(currRef => {
+    // If we've already processed this $ref don't send us into an infinite loop.
+    if ($refs.has(currRef)) {
+      return;
+    }
+
+    $refs.add(currRef);
+    accumulateUsedRefs(schema, $refs, currRef);
+  });
+}
+
+/**
+ * With an array of tags or object of paths+method combinations, reduce an OpenAPI definition to a
+ * new definition that just contains those tags or path + methods.
+ *
+ * @example <caption>Reduce by an array of tags only.</caption>
+ * { APIDEFINITION, { tags: ['pet] } }
+ *
+ * @example <caption>Reduce by a specific path and methods.</caption>
+ * { APIDEFINITION, { paths: { '/pet': ['get', 'post'] } } }
+ *
+ * @example <caption>Reduce by a specific path and all methods it has.</caption>
+ * { APIDEFINITION, { paths: { '/pet': '*' } } }
+ *
+ * @param definition A valid OpenAPI 3.x definition
+ * @param opts Option configuration to reduce by. See the README for details.
```

Comment:
Sorry one more small docs-related comment—is this "See the README for details." comment still applicable?

Reply:
Oh no, I'll remove that.